docs/src/main/asciidoc/cdi.adoc

```diff
@@ -0,0 +1,361 @@
+<2> This callback is invoked before the bean instance is destroyed. It is safe to perform some cleanup tasks here.
+
+TIP: It's a good practice to keep the logic in the callbacks "without side effects", i.e. you should avoid calling other beans inside the callbacks.
+
+=== Interceptors
+
+Interceptors are used to separate cross-cutting concerns from business logic. 
+There is a separate specification - Java Interceptors - that defines the basic programming model and semantics.
+
+.Simple Interceptor Example
+[source,java]
+----
+import javax.interceptor.Interceptor;
+import javax.annotation.Priority;
+
+@Logged <1>
+@Priority(2020) <2>
+@Interceptor <3>
+public class LoggingInterceptor {
+
+   @Inject <4>
+   Logger logger;
+
+   @AroundInvoke <5>
+   Objec logInvocation(InvocationContext context) {
+      // ...log before
+      Objec ret = context.proceed(); <6>
+      // ...log after
+      return ret;
+   }
+    
+}
+----
+<1> This is an interceptor binding annotation that is used to bind our interceptor to a bean. Simply annotate a bean class with `@Logged`.
+<2> `Priority` enables the interceptor and affects the the interceptor ordering. 
+<3> Marks an interceptor component.
+<4> An interceptor instance may be the target of dependency injection.
+<5> `AroundInvoke` denotes a method that interposes on business methods.
+<6> Proceed to the next interceptor in the interceptor chain or invoke the intercepted business method.
+
+NOTE: Instances of interceptors are dependent objects of the bean instance they intercept, i.e. a new interceptor instance is created for each intercepred bean.
+
+=== Events and Observers
+	
+Beans may also produce and consume events to interact in a completely decoupled fashion.
+Any Java object can serve as an event payload.
+The optional qualifiers act as topic selectors.
+
+.Simple Event Example
+[source,java]
+----
+
+class TaskCompleted {
+  // ...
+}
+
+@ApplicationScoped
+class ComplicatedService {
+
+   @Inject
+   Event<Task> event; <1>
+
+   void doSomething() {
+      // ...
+      event.fire(new TaskCompleted()); <2>
+   }
+
+}
+
+@ApplicationScoped
+class Logger {
+
+   void onTaskCompleted(@Observes TaskCompleted task) { <3>
+      // ...log the task
+   }
+
+}
+----
+<1> `javax.enterprise.event.Event` is used to fire events.
+<2> Fire the event synchronously.
+<3> This method is notified when a `TaskCompleted` event is fired.
+
+== Conclusion
+
+CDI is very mature and powerful technology.
+In this guide, we've covered some of the basic topics.
+There is a Quarkus link:cdi-reference[CDI Reference Guide] that covers some Quarkus-specific features and limitations.
+We also recommend you to read the http://docs.jboss.org/cdi/spec/2.0/cdi-spec.html[CDI specification] and the https://docs.jboss.org/weld/reference/latest/en-US/html/[Weld documentation] (Weld is a CDI Reference Implementation) to get acquainted with more complex topics.
```

Comment:
I am not sure linking to Weld makes sense here, I'd probably remove it.

Reply:
The Weld docs are a good source of info about CDI. I'm +10 to keep the link. BTW there are quite a few links to the Weld docs in the guide anyway....